src/main/java/hudson/remoting/Engine.java

```diff
@@ -302,12 +302,24 @@ public void run() {
                 Channel channel = null;
 
                 // Try available protocols.
+                boolean triedAtLeastOneProtocol = false;
                 for (JnlpProtocol protocol : protocols) {
+                    if (!protocol.isEnabled()) {
+                        events.status("Protocol " + protocol.getName() + " is not enabled, skipping");
+                        continue;
+                    }
+                    triedAtLeastOneProtocol = true;
                     events.status("Trying protocol: " + protocol.getName());
                     try {
                         channel = protocol.establishChannel(jnlpSocket, channelBuilder);
                     } catch (IOException ioe) {
-                        events.status("Protocol failed to establish channel", ioe);
+                        events.status("Protocol " + protocol.getName() + " failed to establish channel", ioe);
+                    } catch (RuntimeException e) {
+                        events.status("Protocol " + protocol.getName() + " encountered a runtime error", e);
+                    } catch (Error e) {
+                        events.status("Protocol " + protocol.getName() + " could not be completed due to an error", e);
```

Comment:
I think it's acceptable. If the error is not handleable, it will just blow up in another place